server/src/main/java/org/elasticsearch/cluster/metadata/DesiredNodes.java

```diff
@@ -45,21 +55,48 @@ public record DesiredNodes(String historyID, long version, List<DesiredNode> nod
         PARSER.declareObjectArray(ConstructingObjectParser.constructorArg(), (p, c) -> DesiredNode.fromXContent(p), NODES_FIELD);
     }
 
-    public DesiredNodes {
+    private final String historyID;
+    private final long version;
+    private final Map<String, DesiredNode> nodes;
+
+    public DesiredNodes(String historyID, long version, Collection<DesiredNode> nodes) {
         assert historyID != null && historyID.isBlank() == false;
         assert version != Long.MIN_VALUE;
         checkForDuplicatedExternalIDs(nodes);
+
+        this.historyID = historyID;
+        this.version = version;
+        this.nodes = toMap(nodes);
+    }
+
+    private DesiredNodes(String historyID, long version, Map<String, DesiredNode> nodes) {
+        assert historyID != null && historyID.isBlank() == false;
+        assert version != Long.MIN_VALUE;
+
+        this.historyID = historyID;
+        this.version = version;
+        this.nodes = nodes;
     }
 
-    public DesiredNodes(StreamInput in) throws IOException {
-        this(in.readString(), in.readLong(), in.readList(DesiredNode::new));
+    public static DesiredNodes readFrom(StreamInput in) throws IOException {
+        final String historyId = in.readString();
+        final long version = in.readLong();
+        if (in.getVersion().onOrAfter(Version.CURRENT)) {
+            return new DesiredNodes(historyId, version, in.readMap(StreamInput::readString, DesiredNode::new));
```

Comment:
I think we should preserve order here too? Or perhaps even switch to an ordered map (`TreeeMap`) to ensure a consistent ordering?

I wonder if we even need to change the serialization here? I think we are sending the `externalId` twice now and might as well send it as a list and then reconstruct it on the receiver side?

Reply:
I was thinking about that too. I would expect that the number of desired nodes would be fairly low in most cases. Let me revert this change.